apache/superset
# fix: datatype tracking issue on virtual dataset

### SUMMARY

Column types not detected when creating Virtual Dataset.

If we create a virtual dataset, and try to create a chart based on it, some column types may not be detected as it was defined in original dataset. It was because some columns had no any non-null values when we run the SQL to create a dataset. 
The previous version was tracking column type of columns from the query result values, and that's why some columns had null type even though it had exact type in original dataset. 

To fix this, we need to track the column type from original dataset, not a SQL query running result value.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

**BEFORE**
![image](https://user-images.githubusercontent.com/39701522/168673931-73577f6d-da20-4179-b797-1468d9441466.png)

**AFTER**:
![Screen Shot 2022-05-16 at 16 01 45](https://user-images.githubusercontent.com/39701522/168673789-edcd62f3-9fd2-4d72-a54d-789f9bdb139a.png)
![Screen Shot 2022-05-16 at 16 03 10](https://user-images.githubusercontent.com/39701522/168673819-13c6ecca-6370-407a-9d81-fcb9027f7072.png)

### TESTING INSTRUCTIONS


1. Create a virtual dataset using this query:
```
SELECT *
FROM public."SQLLab_Test_For_VirtualDS"
WHERE site_slug in ('connectmiles', 'rocketmiles')
AND reward_program_slug in ('connectmiles')
LIMIT 10
```

2. Try to create a chart based on virtual dataset.
3. See if all columns have exact data type, not NULL.

Dataset for virtual dataset.
[sqllab_query_publicconnectmiles_all_reservation_stats_20220412T210849.csv](https://github.com/apache/superset/files/8703159/sqllab_query_publicconnectmiles_all_reservation_stats_20220412T210849.csv)



### ADDITIONAL INFORMATION
